src/Autofac/Diagnostics/DiagnosticTracerBase.cs

```diff
@@ -0,0 +1,367 @@
+// Copyright (c) Autofac Project. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Autofac.Diagnostics
+{
+    /// <summary>
+    /// Base class for creating diagnostic tracers that follow Autofac diagnostic events.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Following events from a <see cref="System.Diagnostics.DiagnosticListener"/>
+    /// involves subscribing to that listener with an <see cref="IObserver{T}"/>
+    /// where the observable is a <see cref="KeyValuePair{K,V}"/> of <see cref="string"/>
+    /// event names and <see cref="object"/> event data.
+    /// </para>
+    /// <para>
+    /// This class helps with that by providing strongly typed event handlers
+    /// pre-wired to convert the event data into the right object type, allowing you
+    /// to focus on the handling logic and not the event/data parsing.
+    /// </para>
+    /// <para>
+    /// While technically you could subscribe to non-Autofac events using this
+    /// observer, it's not a general purpose mechanism - it's very much tailored
+    /// to Autofac.
+    /// </para>
+    /// </remarks>
+    /// <seealso cref="DefaultDiagnosticTracer"/>
+    public abstract class DiagnosticTracerBase : IObserver<KeyValuePair<string, object>>
+    {
+        /// <summary>
+        /// The list of event names to which this observer is subscribed.
+        /// </summary>
+        private readonly List<string> _subscriptions = new List<string>();
```

Comment:
Just thought I'd mention that I decided to benchmark List vs HashSet for the size of set we're working with here, just to check; I kinda assumed that a list would be faster for small sets, which it sort of is when searching the beginning of the List, but not the end.

List is wayyy faster when it's empty though, which is our most common case.

|             Method |       Mean |     Error |    StdDev |
|------------------- |-----------:|----------:|----------:|
|          ListEmpty |  0.3399 ns | 0.0134 ns | 0.0119 ns |
|     ListFirstAdded | 13.8114 ns | 0.1138 ns | 0.1008 ns |
|    ListMiddleAdded | 26.4125 ns | 0.1709 ns | 0.1598 ns |
|      ListLastAdded | 53.4440 ns | 0.2396 ns | 0.2242 ns |
|       HashSetEmpty |  2.3426 ns | 0.0317 ns | 0.0265 ns |
|  HashSetFirstAdded | 38.4255 ns | 0.2956 ns | 0.2765 ns |
| HashSetMiddleAdded | 34.3019 ns | 0.1730 ns | 0.1350 ns |
|   HashSetLastAdded | 34.3787 ns | 0.1868 ns | 0.1656 ns |

Reply:
Hmmm, interesting. Keep in mind it's the `DiagnosticListener` / event provider that determines if `IsEnabled` comes out right by iterating across all subscribers (our tracers).

Which is to say, either there _aren't any subscribers_, so it's never hitting our implementation of `IsEnabled` anyway; or there _are subscribers_ (one of our tracers? some other `IObservable<KVP<>>` implementation?) and the subscribers are likely subscribed to at least one event.

I don't think the majority case for our tracers actually will be that they're attached but have no subscriptions. I do think the majority case in general is that the container's `DiagnosticListener` instance won't have any subscribers.

I did [just find that `DiagnosticListener` can do a shortcut call with `IsEnabled()`](https://github.com/dotnet/runtime/blob/e0d20d37592d52d672ff89d5838d0e77e7ee580e/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticListener.cs#L225) such that if there aren't any subscribers then it's a fast null check. Let me add that to all the events, should speed things up.

Reply:
Just pushed that. Benchmark the crap out if it, it should still basically boil down to a check for `subscriptions != null` before doing anything, which should be just as fast as what we had before. Granted, it's going to do that null check for each event rather than just one time, but that still should be pretty fast.